docs/spec/relayer/Packets.md

```diff
@@ -0,0 +1,267 @@
+The value is being verified using the header's app hash that is provided by the corresponding light client.
+We now show the pseudocode for one of those functions:
+
+```go
+GetChannel(chain Chain, 
+           portId Identifier, 
+           channelId Identifier,  
+           proofHeight Height) (ChannelEnd, CommitmentProof) {
+
+    // Query provable store exposed by the full node of chain. 
+    // The path for the channel end is at channelEnds/ports/{portId}/channels/{channelId}".
+    // The membership proof returned is read at height proofHeight. 
+    channel, proof = QueryChannel(chain, portId, channelId, proofHeight) 
+    if proof == nil return { (nil, nil) }
+    
+    header = GetHeader(chain.lc, proofHeight) // get header for height proofHeight using light client of the given chain
+    
+    // verify membership of the channel at path channelEnds/ports/{portId}/channels/{channelId} using 
+    // the root hash header.AppHash
+    if verifyMembership(header.AppHash, proofHeight, proof, channelPath(portId, channelId), channel) {
+        return channel, proof
+    } else { return (nil, nil) }
+}
+```
+If LATEST_HEIGHT is passed as a parameter, the data should be read (and the corresponding proof created) 
+at the most recent height. 
+
+
+## Computing destination chain
+
+```golang
+func GetDestinationInfo(ev IBCEvent, chainA Chain) Chain {
+    switch ev.type {
+        case SendPacketEvent: 
+            channel, proof = GetChannel(chain, ev.sourcePort, ev.sourceChannel, ev.Height)
+            if proof == nil return nil
+                
+            connectionId = channel.connectionHops[0]
+            connection, proof = GetConnection(chain, connectionId, ev.Height) 
+            if proof == nil return nil
+                
+            clientState = GetClientState(chain, connection.clientIdentifier, ev.Height) 
+            return getHostInfo(clientState.chainID) 
+        ...    
+    }
+}
+```
+
+## Datagram creation logic
+
+### PacketRecv datagram creation
+
+```golang
+func createPacketRecvDatagram(ev SendPacketEvent, chainA Chain, chainB Chain, installedHeight Height) PacketRecv {        
+    
+    // Stage 1 
+    // Verify if packet is committed to chain A and it is still pending (commitment exists)
+    
+    proofHeight = installedHeight - 1
+    packetCommitment, packetCommitmentProof = 
+        GetPacketCommitment(chainA, ev.sourcePort, ev.sourceChannel, ev.sequence, proofHeight)     
+    if packetCommitmentProof != nil { return nil }
+        
+    if packetCommitment == null OR
+       packetCommitment != hash(concat(ev.data, ev.timeoutHeight, ev.timeoutTimestamp)) { return nil }
+            
+    // Stage 2 
+    // Execute checks IBC handler on chainB will execute
+    
+    channel, proof = GetChannel(chainB, ev.destPort, ev.destChannel, LATEST_HEIGHT)
+    if proof != nil { return nil }
+    
+    if channel == null OR
+       channel.state != OPEN OR
```

Comment:
I believe on this one we need to retry the handling of the event in order to deal with the opportunistic channel opening case. We have a similar case when we handle the channel related events and where we check if the associated connection is opened.

Reply:
This is a very good point. I assume that we will need to deal with this at a different level; i.e., at the caller level. Even handler should probably just provide more information in a return value (we should return errors with semantics instead of nil) so caller can decide if event should be dropped or it should be retried later.